File: apps/ios/Pebbles/Features/Path/ComposerDraftCoordinator.swift
import Foundation
import Observation
import Supabase
import os

/// Owns the composer's draft lifecycle for whichever composer is open — the
/// server draft (a deliberate "save as draft") and the local snapshot (crash
/// insurance). Extracted from `CreatePebbleSheet+Drafts` (D8).
///
/// The two are one payload shape and nothing else; see
/// `docs/superpowers/specs/2026-07-29-drafts-and-autosave-design.md`.
///
/// This type never owns the draft itself — the sheet holds it in `@State`, the
/// flow holds it on `RecordFlowModel`. It returns instructions and the caller
/// applies them, which is what lets one implementation serve both.
@Observable
@MainActor
final class ComposerDraftCoordinator {

    /// What the composer should do when it opens.
    enum Hydration: Equatable {
        /// Resuming a server draft — hydrate from this payload.
        case resume(PebbleDraftPayload)
        /// A local snapshot exists and is worth offering.
        case offerRestore(PebbleDraftPayload)
        /// Nothing to restore.
        case fresh
    }

    /// Whether a resumed draft's glyph is still usable.
    enum GlyphVerdict: Equatable {
        /// Keep it. Carries the glyph row when we could load it for the preview.
        case usable(Glyph?)
        /// Drop it — the user can no longer attach it.
        case unusable
        /// Verification itself failed; leave the draft alone.
        case unknown
    }

    /// The server draft this composer is bound to — the resumed one, or the one
    /// created by the first "Save as draft".
    private(set) var serverDraftId: UUID?
    private(set) var isSavingDraft = false
    private(set) var hasChecked = false

    /// Non-nil while a restore is on offer.
    private(set) var restorableSnapshot: PebbleDraftPayload?
    var isRestorePromptPresented = false

    private let client: SupabaseClient
    private let drafts: PebbleDraftsService
    private let snapshots: ComposerSnapshotStore
    private let autosave: ComposerAutosave
    private let logger = Logger(subsystem: "app.pbbls.ios", category: "composer-drafts")

    init(client: SupabaseClient, drafts: PebbleDraftsService, snapshots: ComposerSnapshotStore) {
        self.client = client
        self.drafts = drafts
        self.snapshots = snapshots
        self.autosave = ComposerAutosave(store: snapshots)
    }

    // MARK: - Hydration

    /// The pure decision. A resumed server draft wins over a local snapshot —
    /// it is the more deliberate of the two, so we never prompt on top of it.
    ///
    /// Returns nil while `refsLoaded` is false: hydrating before reference data
    /// arrives sanitizes the draft against empty sets and silently drops every
    /// soul and collection (#647), the exact failure M47's D7 exists to prevent.
    /// `nonisolated` because it touches no instance state — being callable
    /// without hopping to the main actor is the point of extracting it.
    nonisolated static func hydration(
        resuming: PebbleDraftRecord?,
        refsLoaded: Bool,
        snapshot: PebbleDraftPayload?
    ) -> Hydration? {
        guard refsLoaded else { return nil }
        if let resuming { return .resume(resuming.payload) }
        if let snapshot, !snapshot.isEmpty { return .offerRestore(snapshot) }
        return .fresh
    }

    /// Applies `hydration(…)` once, recording its side effects. Returns nil
    /// while reference data is loading or when it has already run, so callers
    /// can drive it from both `.task` and an `onChange(of: refs.hasLoaded)`.
    func hydrate(resuming: PebbleDraftRecord?, refsLoaded: Bool) -> Hydration? {
        guard !hasChecked else { return nil }
        guard let decision = Self.hydration(
            resuming: resuming, refsLoaded: refsLoaded, snapshot: snapshots.load()
        ) else { return nil }

        hasChecked = true
        switch decision {
        case .resume:
            serverDraftId = resuming?.id
        case .offerRestore(let snapshot):
            restorableSnapshot = snapshot
            isRestorePromptPresented = true
        case .fresh:
            break
        }
        return decision
    }

    /// The user accepted the restore. Hands back the payload and clears the offer.
    func takeRestorableSnapshot() -> PebbleDraftPayload? {
        defer { restorableSnapshot = nil }
        return restorableSnapshot
    }

    /// The user chose to start fresh.
    func discardSnapshot() {
        restorableSnapshot = nil
        autosave.clear()
    }

    // MARK: - Autosave

    func schedule(_ payload: PebbleDraftPayload) { autosave.schedule(payload) }
    func flush() { autosave.flush() }

    // MARK: - Glyph verification

    /// `can_use_glyph` is what `create_pebble` enforces, so passing here means
    /// publish cannot 42501 on the glyph (M47 D7).
    ///
    /// A verification failure is not a reason to lose the user's glyph:
    /// `.unknown` leaves the draft alone, and publishing will surface a clear
    /// message if it really is unusable.
    func verifyGlyph(glyphId: UUID, userId: UUID) async -> GlyphVerdict {
        do {
            let usable: Bool = try await client
                .rpc(
                    "can_use_glyph",
                    params: ["p_glyph_id": glyphId.uuidString, "p_user": userId.uuidString]
                )
                .execute()
                .value
            guard usable else {
                logger.notice("resumed draft referenced an unusable glyph — dropping it")
                return .unusable
            }
            let glyphs: [Glyph] = try await client
                .from("glyphs")
                .select("id, name, strokes, view_box, user_id")
                .eq("id", value: glyphId)
                .limit(1)
                .execute()
                .value
            return .usable(glyphs.first)
        } catch {
            logger.error("glyph verification failed: \(error.localizedDescription, privacy: .private)")
            return .unknown
        }
    }

    // MARK: - Server draft

    /// Intentional "save as draft". Returns nil on success, or a user-facing
    /// message on failure.
    ///
    /// Deliberately does NOT run `snaps.cancelAndCleanup` — that would delete
    /// the snap the draft references.
    func saveAsDraft(payload: PebbleDraftPayload, userId: UUID) async -> String? {
        isSavingDraft = true
        defer { isSavingDraft = false }
        do {
            serverDraftId = try await drafts.save(payload: payload, id: serverDraftId, userId: userId)
            // Once the draft is on the server the local snapshot is redundant.
            autosave.clear()
            await drafts.refreshCount()
            return nil
        } catch {
            logger.error("save draft failed: \(error.localizedDescription, privacy: .private)")
            return "Couldn't save that draft. Please try again."
        }
    }

    /// Publishing consumed the draft. Runs on soft success too: a 5xx carrying
    /// a `pebble_id` still created the pebble, so a kept draft would duplicate it.
    func consumeAfterPublish() async {
        autosave.clear()
        guard let serverDraftId else { return }
        await drafts.deleteIgnoringFailure(id: serverDraftId)
        self.serverDraftId = nil
        await drafts.refreshCount()
    }
}

File: apps/android/app/src/main/kotlin/app/pbbls/android/services/ComposerDraftCoordinator.kt
package app.pbbls.android.services

import android.util.Log
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import app.pbbls.android.R
import app.pbbls.android.features.path.models.PebbleDraftPayload

private const val TAG = "composer-drafts"

/**
 * The M47 draft lifecycle for a composer, in one object — ports iOS
 * `ComposerDraftCoordinator` (M58 design D8).
 *
 * The riskiest logic around a composer is not the form: it is the draft glue —
 * hydrate-or-offer-restore gated on `refs.hasLoaded` (#647), the `can_use_glyph`
 * check (M47 D7), save-as-draft, and consuming the draft on the soft-success
 * path so a kept draft cannot duplicate a published pebble. Every one of those
 * is a bug already found and fixed once, and the failure mode of a second copy
 * is silent: a new composer keeps working while quietly losing the M47 fixes.
 *
 * `CreatePebbleScreen` still carries its own inline copy of this glue —
 * migrating it here is a follow-up, not part of the flow's landing, because
 * touching a shipped composer is a refactor of working code.
 */
class ComposerDraftCoordinator(
    private val drafts: PebbleDraftsService,
    private val snapshots: ComposerSnapshotStore,
) {
    /** What [hydrate] decided the composer should open with. */
    sealed interface Decision {
        /** A server draft was passed in: hydrate from its payload. */
        data class Resume(
            val payload: PebbleDraftPayload,
        ) : Decision

        /** A local crash snapshot exists: the composer should prompt. */
        data object OfferRestore : Decision

        /** Nothing to restore. */
        data object Fresh : Decision
    }

    /** Outcome of the `can_use_glyph` check on a resumed draft's glyph. */
    sealed interface GlyphVerdict {
        data object Usable : GlyphVerdict

        data object Unusable : GlyphVerdict

        /** The check itself failed — keep the glyph; publish will say if it really is unusable. */
        data object Unknown : GlyphVerdict
    }

    private val autosave = ComposerAutosave(snapshots.asSink())

    /** The row this composer is bound to — the resumed one, or the first save's. */
    var serverDraftId: String? = null
        private set

    /** True while the restore prompt is up, so autosave does not overwrite the pending answer. */
    var isRestorePromptPresented: Boolean by mutableStateOf(false)
        private set

    private var restorable: PebbleDraftPayload? = null
    private var hasHydrated = false

    /**
     * Decide what the composer opens with. Returns null until reference data has
     * loaded, and only ever decides once.
     *
     * Gated on [refsLoaded] because hydrating before the souls / collections
     * caches arrive would sanitize against empty sets and silently drop every
     * soul and collection on the draft (#647). Resuming a server draft wins over
     * the local snapshot — it is the more deliberate of the two, so we never
     * prompt on top of it.
     */
    fun hydrate(
        resuming: PebbleDraftRecord?,
        refsLoaded: Boolean,
    ): Decision? {
        if (hasHydrated || !refsLoaded) return null
        hasHydrated = true
        if (resuming != null) {
            serverDraftId = resuming.id
            return Decision.Resume(resuming.payload)
        }
        restorable = snapshots.load()
        return if (restorable != null) {
            isRestorePromptPresented = true
            Decision.OfferRestore
        } else {
            Decision.Fresh
        }
    }

    /** The user accepted the restore prompt. Consumes the snapshot. */
    fun takeRestorableSnapshot(): PebbleDraftPayload? {
        isRestorePromptPresented = false
        val snapshot = restorable
        restorable = null
        return snapshot
    }

    /** The user declined the restore prompt, or is leaving without keeping anything. */
    fun discardSnapshot() {
        isRestorePromptPresented = false
        restorable = null
        autosave.clear()
    }

    /** Remember the current payload; the caller owns the debounce delay. */
    fun stage(payload: PebbleDraftPayload) {
        autosave.stage(payload)
    }

    /** Write the staged snapshot — after the debounce, or when leaving the foreground. */
    fun flush() {
        autosave.flush()
    }

    /**
     * Intentional "save as draft". Returns a message resource id on failure, or
     * null on success.
     *
     * Deliberately does NOT clean up the attached snap: that would delete from
     * Storage the very object the draft references (M47 D3).
     */
    suspend fun saveAsDraft(
        payload: PebbleDraftPayload,
        userId: String,
    ): Int? =
        try {
            serverDraftId = drafts.save(payload = payload, id = serverDraftId, userId = userId)
            // Once the draft is on the server the local snapshot is redundant.
            autosave.clear()
            null
        } catch (e: Exception) {
            Log.e(TAG, "save draft failed", e)
            R.string.draft_save_error
        }

    /**
     * Publishing consumed the draft. Runs on the soft-success path too: a 5xx
     * carrying a `pebble_id` still created the pebble, so leaving the draft would
     * duplicate it in the list. A failed delete must not fail the publish.
     */
    suspend fun consumeAfterPublish() {
        autosave.clear()
        val id = serverDraftId ?: return
        try {
            drafts.delete(id)
        } catch (e: Exception) {
            Log.w(TAG, "draft cleanup after publish failed", e)
        }
        serverDraftId = null
    }

    /**
     * Whether a resumed draft's glyph is still usable. `can_use_glyph` is the
     * predicate `create_pebble` enforces, so checking it while hydrating means
     * publishing cannot fail on 42501 later (M47 D7).
     */
    suspend fun verifyGlyph(
        glyphId: String,
        userId: String,
    ): GlyphVerdict =
        try {
            if (drafts.canUseGlyph(glyphId, userId)) {
                GlyphVerdict.Usable
            } else {
                Log.i(TAG, "resumed draft referenced an unusable glyph — dropping it")
                GlyphVerdict.Unusable
            }
        } catch (e: Exception) {
            Log.e(TAG, "glyph verification failed", e)
            GlyphVerdict.Unknown
        }
}
